chainextensions/dapps-staking/impl/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+            DappsStakingFunc::EraRewards => {
+                let arg: u32 = env.read_as()?;
+                let era_info = pallet_dapps_staking::GeneralEraInfo::<T>::get(arg)
+                    .ok_or(DispatchError::Other("general_era_info call failed"));
+                let reward = era_info.map_or(Zero::zero(), |r| {
+                    r.rewards.stakers.saturating_add(r.rewards.dapps)
+                });
+                env.write(&reward.encode(), false, None).map_err(|_| {
+                    DispatchError::Other(
+                        "[ChainExtension] DappsStakingExtension failed to write result",
+                    )
+                })?;
+            }
+
+            // DappsStaking - read_era_staked()
+            DappsStakingFunc::EraStaked => {
+                let arg: u32 = env.read_as()?;
+                let era_info = pallet_dapps_staking::GeneralEraInfo::<T>::get(arg)
+                    .ok_or(DispatchError::Other("general_era_info call failed"));
+                let staked_amount = era_info.map_or(Zero::zero(), |r| r.staked);
+                env.write(&staked_amount.encode(), false, None)
+                    .map_err(|_| {
+                        DispatchError::Other(
+                            "[ChainExtension] DappsStakingExtension failed to write result",
+                        )
+                    })?;
+            }
+
+            // DappsStaking - read_staked_amount()
+            DappsStakingFunc::StakedAmount => {
+                let staker: T::AccountId = env.read_as()?;
+                let ledger = pallet_dapps_staking::Ledger::<T>::get(&staker);
+                env.write(&ledger.locked.encode(), false, None)
+                    .map_err(|_| {
+                        DispatchError::Other(
+                            "[ChainExtension] DappsStakingExtension failed to write result",
+                        )
+                    })?;
+            }
+
+            DappsStakingFunc::StakedAmountOnContract => {
+                let args: DappsStakingAccountInput = env.read_as()?;
+                let staker = T::AccountId::decode(&mut args.staker.as_ref()).unwrap();
+                let contract = Self::decode_smart_contract(args.contract)?;
+                let staking_info =
+                    pallet_dapps_staking::GeneralStakerInfo::<T>::get(&staker, &contract);
+                let staked_amount = staking_info.latest_staked_value();
+                env.write(&staked_amount.encode(), false, None)
+                    .map_err(|_| {
+                        DispatchError::Other(
+                            "[ChainExtension] DappsStakingExtension failed to write result",
+                        )
+                    })?;
+            }
+
+            // DappsStaking - read_contract_stake()
+            DappsStakingFunc::ReadContractStake => {
+                let contract_bytes: [u8; 32] = env.read_as()?;
+                let contract = Self::decode_smart_contract(contract_bytes)?;
```

Comment:
This is somehow limiting - what if WASM contract wants to access EVM contract?
Since we're aiming to be a X-VM chain, isn't it important to consider that?

Reply:
@Dinonard I'll do it in another PR as there is a lot of changes to do this.
I need to pass the SmartCotnract enum instead of a 32 bytes array

Reply:
Sure, let's just make sure we don't deploy on Shiden before that is deployed & tested on Shibuya :+1: